### PR TITLE
KAFKA-6958: Overload methods for group and windowed stream to allow to name operation name using the new Named class

### PR DESCRIPTION
**Pull request by fhussonnois**
Hi @mjsax @bbejeck

This is the last PR for the KIP-307.

NOTE : [PR 6412](https://github.com/apache/kafka/pull/6412) should be merge first

Thanks a lot for the review.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java]: This should not have the `sink` suffix

**Review by bbejeck — COMMENTED** [review]
Made a pass over `SessionedXX` `TimeWindowed` and aggregate builders.  Just one minor comment otherwise looks reasonable.

**Comment by bbejeck** [comment]
@fhussonnois please rebase this PR

**Comment by bbejeck** [comment]
ping @vvcephei @ableegoldman for reviews

**Comment by bbejeck** [comment]
@fhussonnois #6412 is merged now, can you rebase this PR and ping us when it's ready for review?

**Comment by fhussonnois** [comment]
@bbejeck This PR has been rebase (sorry I just remember I've forgotten to ping you) thank you

**Comment by bbejeck** [comment]
>This PR has been rebase (sorry I just remember I've forgotten to ping you) thank you

@fhussonnois no worries! Thanks for the rebase, I'll make a pass on this last PR soon

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
@fhussonnois the build failures are related, the changes need to go into the scala API as well - `Execution failed for task ':streams:streams-scala:compileTestScala'`

**Comment by bbejeck** [comment]
Java 8 and Java 11 Scala 2.12 failed with
``` 
      fatal: unable to connect to github.com:
      github.com: Temporary failure in name resolution
```
Java 11 2.13 failed with `SaslScramSslEndToEndAuthorizationTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl`

retest this please

**Comment by bbejeck** [comment]
Test results already removed.

retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java]: nit: add `a {@link Named} config`

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java]: nit: add `a {@link Named} config`

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java]: nit add `a {@link Named} config`

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/SessionWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/TimeWindowedKStream.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java]: shouldn't the `table-source` suffix go with the `sourceName`?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java]: should this have a suffix at all?  I don't think the `funcName` requires one.

If it does, at a minimum we should not use `-sink` as it's used above and doesn't accurately describe its role.

**Review by bbejeck — COMMENTED** [review]
@fhussonnois thanks for your persistence on KIP-307!  I've made a pass over this final PR and overall it looks good, I only have minor comments.

**Comment by bbejeck** [comment]
Java 11, Scala 2.12 failed, Java 8 and Java 11 Scala 2.11 passed, test results already cleaned up.

retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java]: @bbejeck ,You're right, it seems OK to remove the suffix from the function name. The user should ensure that a unique name is given to each node. So, I've removed the suffix and I also add a unit test.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (fhussonnois) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java]: @bbejeck, Yes, that seems more logical. I made the modification for the global table too. Also, I think the suffix "-source" is sufficient.

**Review by bbejeck — APPROVED** [review]
Thanks for the update @fhussonnois LGTM.

**Comment by bbejeck** [comment]
Java 8/2.11 and Java 11/2.13 failed
Java 11/2.12 passed   
test results already cleaned out

retest this please

**Comment by bbejeck** [comment]
ping @guozhangwang, @mjsax, @vvcephei, @ableegoldman, @cadonna, @abbccdda 
for a second review

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:164]: Would this result in a different name for the source than the prior code? (Not sure if it matters...)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java:76]: did you want to add a null check for this as well (like you did for `aggregate`)?

File: streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala
Comment (vvcephei) [inline streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala]: Were these changes necessary to disambiguate the methods? It would be important to know, because it would constitute an API-breaking change in the Scala DSL.

Comment (vvcephei) [reply streams/streams-scala/src/test/scala/org/apache/kafka/streams/scala/TopologyTest.scala]: Ok, I've just confirmed that this _is_ a breaking change in the Scala DSL. I'm not sure why but introducing the `KGroupedStream#aggregate(Initializer, Aggregator, Named)` method creates ambiguity about how the type system should infer the generic parameters to Materialized here.

This also resolves the ambiguity:

```scala
  val stream2: KStreamJ[String, Integer] = mappedStream.groupByKey
        .aggregate(initializer, aggregator, MaterializedJ.`with`(Serdes.String, SerdesJ.Integer): MaterializedJ[String, Integer, KeyValueStore[Bytes, Array[Byte]]])
        .toStream
```

Either way, it's a breaking change, which we should probably not merge.

One simple fix is simply not to add that particular overload. The full expression with Named _and_ Materialized would still be present.

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:454]: It seems weird that these changed. Is that expected and ok?

**Review by vvcephei — CHANGES_REQUESTED** [review]
Hey @fhussonnois , thanks for this latest edition in the ongoing saga of adding Named :)

This pretty much looks good to me, I just had a few comments and questions.

The big concern is the Scala DSL. It seems like this is actually a significant problem.

Thanks,
-John

**Comment by bbejeck** [comment]
@fhussonnois regarding the overloads 
`KGroupedStream#aggregate(initializer, aggregator, named)` and `KGroupedStream#reduce(reducer, named)` let's go ahead and just drop those for now and we can figure out how to work them in later on.

The KIP will need to get updated with this information as well.

**Comment by bbejeck** [comment]
Java 8/2.11, and Java 11/2.13 passed
Java 11/2.12 failed `Build timed out (after 240 minutes). Marking the build as aborted.`

retest this please

**Comment by bbejeck** [comment]
Java 11/2.13 failed, test results already gone

retest this please

File: streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/StreamsBuilderTest.java:454]: Hey @fhussonnois or @bbejeck , what do you think about this?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:164]: Hey @fhussonnois or @bbejeck , what do you think about this?

**Review by vvcephei — APPROVED** [review]
This LGTM now. I had one unanswered question, but I don't think we need to block on it. We can always do a follow-up patch if it's important.

Thanks so much @fhussonnois , for all your work on this!

**Comment by bbejeck** [comment]
Merged #6413 into trunk.

**Comment by bbejeck** [comment]
Thanks for all the hard work @fhussonnois !!

**Comment by fhussonnois** [comment]
Thank you very much @bbejeck for you help on that KIP.

**Comment by mjsax** [comment]
Congrats @fhussonnois! Great work!

**Pull request merged**
